libsolidity/codegen/ABIFunctions.cpp

```diff
@@ -319,7 +322,56 @@ string ABIFunctions::cleanupFunction(Type const& _type, bool _revertOnFailure)
 				StateMutability::Payable :
 				StateMutability::NonPayable
 			);
-			templ("body", "cleaned := " + cleanupFunction(addressType, _revertOnFailure) + "(value)");
+			templ("body", "cleaned := " + cleanupFunction(addressType) + "(value)");
+			break;
+		}
+		case Type::Category::Enum:
+		{
+			// Out of range enums cannot be truncated unambigiously and therefore it should be an error.
+			templ("body", "cleaned := " + validatorFunction(_type) + "(value)");
+			break;
+		}
+		case Type::Category::InaccessibleDynamic:
+			templ("body", "cleaned := 0");
+			break;
+		default:
+			solAssert(false, "Cleanup of type " + _type.identifier() + " requested.");
+		}
+
+		return templ.render();
+	});
+}
+
+string ABIFunctions::validatorFunction(Type const& _type, bool _revertOnFailure)
+{
+	string functionName = string("validator_") + (_revertOnFailure ? "revert_" : "assert_") + _type.identifier();
+	return createFunction(functionName, [&]() {
+		Whiskers templ(R"(
+			function <functionName>(value) -> cleaned {
+				<body>
+			}
+		)");
+		templ("functionName", functionName);
+		switch (_type.category())
+		{
+		case Type::Category::Address:
+		case Type::Category::Integer:
+		case Type::Category::RationalNumber:
+		case Type::Category::Bool:
+		case Type::Category::FixedPoint:
+		case Type::Category::Function:
+		case Type::Category::Array:
+		case Type::Category::Struct:
+		case Type::Category::FixedBytes:
+		case Type::Category::Contract:
+		{
+			Whiskers w("cleaned := <cleanup>(value) if iszero(eq(cleaned, value)) { <failure> }");
```

Comment:
Is this really the cheapest way to do this? Or more like a working placeholder for later improvement?
E.g. for unsigned integers it'd just be ``if iszero(iszero(and(value, 0xFF...FF00...00))) { <failure> }`` and I'm not sure optimization will get us there without major hoops...

Reply:
Actually: should this function return anything? I'd imagine the optimizer would have a better chance, if ``cleaned`` was a local variable... So in order not to have to refactor too much to that end - do you think restructuring as follows might help the optimizer:
```
if iszero(eq(<cleanup>(value), value)) { <failure> }
cleaned := value
```

Reply:
Otherwise we'd probably end up with something along the lines of:
```
function cleanup_uint128(value) -> cleaned {
  cleaned := and(value,0x00..00FF..FF)
  if iszero(eq(cleaned, value)) { <failure> }
}
```
Hm... that's not particularly worse than the other stuff - maybe I haven't thought this through sufficiently... it should be fine for now in any case.

Reply:
Especially since I'm basically repeating the discussion you had earlier in the comments above :-). Still I'd think about ``if iszero(eq(<cleanup>(value), value)) { <failure> } cleaned := value`` - it's equally simple and I could imagine it to be quite beneficial for optimization, especially when inlined.

Reply:
@axic initially wrote a different function, but I thought it might be better to have a function we are sure is correct and let the optimizer take care of it. I also suggested that the validator should not return anything, but it is kind of easier to use like that, because you do not have to introduce another local vriable.

Reply:
The optimizer will mangle this through SSA anway, so it should not make a big difference.

Reply:
I'm still not sure whether it's better to return the cleaned value or the original argument (they are the same, but so far I don't think the optimizer will be able to use that fact) - that'll make a difference even with SSA, but I'm not sure which is better and we can change it later, so it's fine.

Reply:
Ah you are right - now I read your comment properly - sorry about that. It might actually be beneficial to return the cleaned, because we have more information about it. If value range tracking is implemented properly, the information should actually transfer to the other value, so I don't really know...

Reply:
I'll change it to return the value for now.

Reply:
Yeah - that's what I meant - currently I think either way to do it can be better depending on the particular case... but yeah - value range tracking will help at some point.